contracts/SimpleGAMultiSig.aes

```diff
@@ -0,0 +1,153 @@
+    check_nonce(n)
+    switch(is_expired())
+      true => 
+        abort("ERROR_TX_EXPIRED")
+      false =>
+        require(has_consensus(), "ERROR_NO_CONSENSUS")
+        switch(Auth.tx_hash)
+          None => abort("ERROR_NO_AUTH_CONTEXT")
+          Some(ga_tx_hash) =>
+            require(ga_tx_hash == Option.force(state.current_tx).ga_tx_hash, "ERROR_UNEQUAL_HASHES")
+            require(is_authorized_and_valid(Call.caller, ga_tx_hash, sig), "ERROR_NOT_AUTHORIZED")
+            Chain.event(TxAuthorized(ga_tx_hash))        
+            true
+    
+  // propose a new tx valid for the given ttl
+  stateful entrypoint propose(ga_tx_hash : hash, ttl : Chain.ttl, sig : signature, n : int) =
+    check_nonce(n)
+    require(is_authorized_and_valid(Call.caller, ga_tx_hash, sig), "ERROR_NOT_AUTHORIZED")
+    switch(state.current_tx)
+      None => set_current_tx(ga_tx_hash, ttl)
+      Some(tx) => 
+        switch(is_expired())
+          false => abort("ERROR_A_TX_IS_ALREADY_PROPOSED")
+          true => set_current_tx(ga_tx_hash, ttl)
+  
+  // signer confirms tx
+  stateful entrypoint confirm(sig : signature, n : int) = 
+    check_nonce(n)
+    switch(state.current_tx)
+      None => abort("ERROR_NOTHING_TO_CONFIRM")
+      Some(tx) =>
+        require(is_authorized_and_valid(Call.caller, tx.ga_tx_hash, sig), "ERROR_NOT_AUTHORIZED")
+        require(!List.contains(Call.caller, tx.confirmed_by),"ERROR_ALREADY_CONFIRMED")
+        require(!is_expired(), "ERROR_TX_ALREADY_EXPIRED")        
+        let updated_tx = tx{confirmed_by=Call.caller :: tx.confirmed_by}
+        put(state{current_tx=Some(updated_tx)})
+        Chain.event(TxConfirmed(tx.ga_tx_hash, Call.caller))
+        if(has_consensus())
+          Chain.event(TxConsensusReached(tx.ga_tx_hash))
+  
+  // revoke the current tx and clean state
+  stateful entrypoint revoke(sig : signature, n: int) =
+    check_nonce(n)
+    switch(state.current_tx)
+      None => abort("ERROR_NOTHING_TO_REVOKE")
+      Some(tx) =>
+        require(is_authorized_and_valid(Call.caller, tx.ga_tx_hash, sig), "ERROR_NOT_AUTHORIZED")
+        put(state{current_tx = None})
+        Chain.event(TxRevoked(Call.caller))
+
+  // check if given address is within the list of multisig_users
+  // AND 
+  // check that the signature of the proposed ga tx was made using the private key of the given address
+  entrypoint is_authorized_and_valid(addr : address, ga_tx_hash_to_verify : hash, sig : signature) : bool =
+    List.contains(addr, state.multisig_users) && Crypto.verify_sig(ga_tx_hash_to_verify, addr, sig)
+
+  // validate if enough signers have confirmed
+  entrypoint has_consensus() : bool = 
+    switch(state.current_tx)
+      None => false
+      Some(tx) => List.length(tx.confirmed_by) >= state.confirmations_required
+	
+  // check if current_tx is already expired
+  function is_expired() : bool =
+    switch(state.current_tx)
+      None => true
```

Comment:
why not abort?

Reply:
agree, abort makes more sense

Reply:
passing the `tx` now based on a previous `Option.force_msg` to the functions that should return `true/false` (has_consensus and is_expired)